doc/advanced/content/executor_design.rst

```diff
@@ -0,0 +1,421 @@
+feature and is user dependent, PCL code cannot provide any guarantee that custom executors will work as
+expected for PCL functions. Make sure to look and understand the code for the function in which
+you are using a custom executor and determine whether the executor will provide the expected results.
+
+**Implementation**
+
+Below we show a simplified version of the inline executor. Look at the Code API for more details.
+
+.. code-block:: cpp
+
+   // Properties are specified as template parameters
+   template <typename Blocking = blocking_t::always_t, typename ProtoAllocator = std::allocator<void>>
+   struct inline_executor {
+
+      // Shape and index are specified to the custom PCL type uindex_t
+      using shape_type = uindex_t ;
+      using index_type = uindex_t ;
+
+
+      // execute invokes the callable exactly once
+      template <typename F>
+      void execute(F&& f) const {
+        f();
+      }
+
+      // bulk_execute invokes the callable n times and passes the index of execution agent
+      // In case of inline_executor each execution agent is mapped to the same thread is invoked sequentially
+      template <typename F, typename... Args>
+      void bulk_execute(F&& f, const shape_type& n) const {
+        for (shape_type index = 0; index <  n; ++index)
+          f(index);
+      }
+
+      // The query customization point being specialized to show that the executor supports
+      // the blocking.always property
+      static constexpr auto query(const blocking_t&) noexcept {
+        return blocking_t::always;
+      }
+
+      // The require customization point being specialized to allow strongly associating the executor with
+      // the blocking.always property
+      inline_executor<blocking_t::always_t, ProtoAllocator> require(const blocking_t::always_t&) const {
+        return {};
+      }
+    };
+
+
+Best Fit Executor
+=================================
+
+In most scenarios users of PCL do not care much about performance and simply want their code to run
+fast without any additional steps performed from their side. For this reason their was a need for
+mechanism called best fit, which automatically selects best possible executor and its customizes its
+property in order to give the user good performance out of the box based on their system configurations
+and the function they are calling.
+
+The mechanism works be choosing an executor based on the availability of an executor (depends on
+hardware/software of a system) and the priority of executors specified in function by PCL
+maintainers & contributors which will give good performance. The executor properties can also be
+customized on for a better fit for certain scenarios. Besides this runtime checks are also be specified
+and on the basis of these runtime checks the executors are further filtered to selected most appropriate
+one. The two supported mechanisms in PCL currently are `enable_exec_with_priority` and
```

Comment:
Revise this sentence for commas, grammar and to reduce the repetition of the expression "runtime checks".